File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:

        res = []
        curr = []

        def dfs(node, curr):
            
            if not node:
                return

            curr.append(node.val)

            if not node.left and not node.right:
                res.append(curr.copy())
            else:

                if node.left:
                    dfs(node.left, curr)
                if node.right:
                    dfs(node.right, curr)

            curr.pop()

        dfs(root, curr)

        result = 0
        for i in res:
            decimalVal = int("".join(map(str, i)), 2)
            result += decimalVal
        
        return result
        

File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumRootToLeaf(TreeNode* root) 
    {
        //Base Case
        if (!root)
        {
            return 0;
        }

        vector<int> curr;
        
        //Grab everything
        dfs(root, curr);

        printMatrix(res);

        int total = 0;

        for (int i = 0; i < res.size(); i++)
        {
            std::string num;
            for (int j = 0; j < res[i].size(); j++)
            {
                num += (res[i][j] + '0');
            }

            total += stoi(num, nullptr, 2);
        }

        return total;
    }

    void printMatrix(const std::vector<std::vector<int>>& matrix)
    {
        for (const auto& row : matrix)
        {
            for (int val : row)
            {
                std::cout << val << " ";
            }
            std::cout << "\n";
        }
    }

    void dfs(TreeNode* node, vector<int> curr)
    {

        int val = node->val;
        curr.push_back(val);

        //Check leaf
        if (!node->left && !node->right)
        {
            res.push_back(curr);
            return;
        }

        
        if (node->left)
        {
            dfs(node->left, curr);
        }
        if (node->right)
        {
            dfs(node->right, curr);
        }

    }

private:
    vector<vector<int>> res;
};